agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java

```diff
@@ -195,6 +205,48 @@ protected void onRequest(final Span span, final REQUEST request) {
     // TODO set resource name from URL.
   }
 
+  protected void onConnectionAndRequest(Span span, CONNECTION connection, REQUEST request) {
+    String flavor = flavor(connection, request);
+    if (flavor != null) {
+      SemanticAttributes.HTTP_FLAVOR.set(span, flavor);
+    }
+    SemanticAttributes.HTTP_CLIENT_IP.set(span, clientIP(connection, request));
+  }
+
+  protected String clientIP(CONNECTION connection, REQUEST request) {
+    // try Forwarded
+    String forwarded = requestHeader(request, "Forwarded");
+    if (forwarded != null) {
+      // may be split by ; or ,
+      for (String forwardedValues : FORWARDED_SPLIT_PATTERN.split(forwarded)) {
+        forwardedValues = forwardedValues.trim();
+        String[] split = EQUALS_PATTERN.split(forwardedValues, 2);
```

Comment:
can you convert `EQUALS_PATTERN` to indexOf/substring to avoid that regex? I don't think that would add code complexity.

replacing `FORWARDED_SPLIT_PATTERN` would definitely add code complexity, so i don't mind keeping

Reply:
Out of curiosity, what do you think about this? I noticed the lines of code are very close to what we have right now even though we define a helper method.

```
int forIndex = forwarded.indexOf("for=");
int endIndex = findForwardedEnd(forwarded, forIndex + 4);
forwarded.substring(forIndex + 4, endIndex);

int find forwardedEnd(String s, int start) {
  for (int i = start; i < s.length(); i++) {
    char c = s.charAt(i);
    if (c == ',' || c == ';') {
      return i;
    }
  }
  return s.length()
```

Reply:
@anuraaga this would work for regular cases but would not catch a malformed header such as "for=, for=1.1.1.1" That's probably okay though. I will use your implementation.